File: chapter01/02/Prog.java
import java.util.function.Function;

// javac Prog.java
// java Prog

public class Prog {
  public static void main(String[] args) {
    Function<String, Integer> length = s -> s.length();
    Function<Integer, Boolean> isEmpty = l -> l == 0;
    var isEmptyString = compose(isEmpty, length);
    System.out.println(isEmptyString.apply(""));
    System.out.println(isEmptyString.apply("nest"));
  }

  public static <A, B, C> Function<A, C> compose(Function<B, C> g, Function<A, B> f) {
    return (A a) -> g.apply(f.apply(a));
  }
}

File: chapter01/02/prog.hs
-- ghc prog.hs -o a.out
-- ./a.out

isEmpty :: Int -> Bool
isEmpty len = len == 0

isEmptyString :: String -> Bool
isEmptyString = isEmpty . length

main :: IO ()
main = do putStrLn (show (isEmptyString ""))
          putStrLn (show (isEmptyString "next"))